src/app/InvokeInteraction.h

```diff
@@ -0,0 +1,438 @@
+/*
+ *
+ *    Copyright (c) 2020-2021 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file defines a set of classes that applications can use to initiate and respond to
+ *      IM commands.
+ */
+
+#pragma once
+
+#include <core/CHIPCore.h>
+#include <messaging/ExchangeContext.h>
+#include <messaging/ExchangeMgr.h>
+#include <messaging/Flags.h>
+#include <protocols/Protocols.h>
+#include <support/BitFlags.h>
+#include <support/CodeUtils.h>
+#include <support/DLLUtil.h>
+#include <support/logging/CHIPLogging.h>
+#include <system/SystemPacketBuffer.h>
+#include <system/TLVPacketBufferBackingStore.h>
+#include "IEncodableElement.h"
+#include <app/MessageDef/CommandDataElement.h>
+#include <app/MessageDef/CommandList.h>
+#include <app/MessageDef/InvokeCommand.h>
+#include <app/util/basic-types.h>
+#include "messaging/ExchangeDelegate.h"
+#include "protocols/secure_channel/Constants.h"
+#include <functional>
+
+namespace chip {
+namespace app {
+
+/*
+ * @brief
+ *
+ * This class encapsulates the data that goes into a status response message.
+ */
+struct StatusResponse {
+public:
+    bool IsError() { return generalCode != Protocols::SecureChannel::GeneralStatusCode::kSuccess; }
+    Protocols::SecureChannel::GeneralStatusCode generalCode = Protocols::SecureChannel::GeneralStatusCode::kSuccess;
+    Protocols::Id protocolId = Protocols::Id(Common, 0);
+    uint16_t protocolCode = 0;
+};
+
+/**
+ * @brief
+ *
+ * Encapsulates parameters that go into the header of an IM Invoke Request or Response message.
+ */
+struct CommandParams
+{
+    enum class TargetType : uint8_t
+    {
+        kTargetEndpoint     = 0x01, /**< Set when the EndpointId field is valid */
+        kTargetGroup        = 0x02, /**< Set when the GroupId field is valid */
+    };
+
+    template <class CodeGenType>
+    CommandParams(CodeGenType& t, chip::EndpointId endpointId, bool expectsResponses = true) {
+        ClusterId = (uint16_t)CodeGenType::GetClusterId();
+        CommandId = (uint8_t)CodeGenType::GetCommandId();
+        EndpointId = endpointId;
+        ExpectsResponses = expectsResponses;
+    }
+
+    CommandParams() { }
+
+    chip::EndpointId EndpointId = 0;
+    chip::GroupId GroupId = 0;
+    chip::ClusterId ClusterId = 0;
```

Comment:
Sync Notes: Not needed since the passed in cluster object will contain a means to extract this.